src/articles/articles.repository.ts

```diff
@@ -0,0 +1,176 @@
+              'favorited',
+            ),
+          favoriteCount: sql<number>`count(*)::integer`.as('favoriteCount'),
+        })
+        .from(articles)
+        .leftJoin(favoriteArticles, eq(favoriteArticles.articleId, articles.id))
+        .leftJoin(users, eq(users.id, favoriteArticles.userId))
+        .where(and(...articleFilters))
+        .groupBy(articles.id),
+    );
+
+    const results = await this.db
+      .with(authorsWithFollowersCTE, articlesWithLikesCTE)
+      .select({
+        slug: articles.slug,
+        title: articles.title,
+        description: articles.description,
+        tagList: articles.tagList,
+        createdAt: articles.createdAt,
+        updatedAt: articles.updatedAt,
+        favorited: articlesWithLikesCTE.favorited,
+        favoritesCount: articlesWithLikesCTE.favoriteCount,
+        author: {
+          username: authorsWithFollowersCTE.authorUsername,
+          bio: authorsWithFollowersCTE.authorBio,
+          image: authorsWithFollowersCTE.authorImage,
+          following: authorsWithFollowersCTE.authorFollowing,
+        },
+      })
+      .from(articles)
+      .innerJoin(
+        articlesWithLikesCTE,
+        eq(articlesWithLikesCTE.articleId, articles.id),
+      )
+      .innerJoin(
+        authorsWithFollowersCTE,
+        eq(authorsWithFollowersCTE.authorId, articles.authorId),
+      )
+      .limit(limit)
+      .offset(offset)
+      .orderBy(desc(articles.createdAt));
+
+    return results;
+  }
+
+  async findBySlug(slug: string) {
+    const result = await this.db.query.articles.findFirst({
+      where: eq(articles.slug, slug),
+      with: {
+        author: {
+          with: {
+            followers: true,
+          },
+        },
+        favoritedBy: true,
+      },
+    });
+    return result;
+  }
+
+  async findById(id: number) {
+    const result = await this.db.query.articles.findFirst({
+      where: eq(articles.id, id),
+      with: {
+        author: {
+          with: {
+            followers: true,
+          },
+        },
+        favoritedBy: true,
+      },
+    });
+    return result;
+  }
+
+  async createArticle(article: ArticleToCreate) {
+    const results = await this.db.insert(articles).values(article).returning();
+    const newArticle = results[0];
+    return this.findById(newArticle.id);
+  }
+
+  async updateArticle(
+    articleId: number,
+    article: ArticleToUpdate,
+    currentUserId: number,
+  ) {
+    const valuesToSet = Object.entries(article).reduce(
+      (acc: { [key: string]: string | string[] | Date }, [key, value]) => {
+        if (value !== undefined) {
+          acc[key] = value;
+        }
+        return acc;
+      },
+      {},
+    );
+    valuesToSet.updatedAt = new Date();
+    await this.db
+      .update(articles)
+      .set(valuesToSet)
+      .where(
+        and(eq(articles.id, articleId), eq(articles.authorId, currentUserId)),
+      );
+  }
```

Comment:
```suggestion

    valuesToSet.updatedAt = new Date();
    await this.db
      .update(articles)
      .set({
        ...shake(article),
        updatedAt: new Date(),
      })
      .where(
        and(eq(articles.id, articleId), eq(articles.authorId, currentUserId)),
      );
  }
```
it's easier to wrap your head around - prefer to avoid `reduce` when possible, because they come with a cognitive overhead.

`shake` from Radashi: https://radashi.js.org/reference/object/shake/

Reply:
Isn't this 'shake' utility just a reduce [wrapped with some TS typing](https://github.com/radashi-org/radashi/blob/main/src/object/shake.ts)? I'm not a big fan of adding dependencies to do something simple.
And if we copy the code of the shake function, then we just made the problem harder - we have more code to maintain and it uses TS generics.

Reply:
Went with a simpler approach